infoschema/cache.go

```diff
@@ -125,14 +155,36 @@ func (h *InfoCache) Insert(is InfoSchema, snapshotTS uint64) bool {
 	defer h.mu.Unlock()
 
 	version := is.SchemaMetaVersion()
-	i := sort.Search(len(h.cache), func(i int) bool {
-		return h.cache[i].SchemaMetaVersion() <= version
-	})
 
-	if h.maxUpdatedSnapshotTS < snapshotTS {
-		h.maxUpdatedSnapshotTS = snapshotTS
+	i := sort.Search(len(h.sortedTS), func(i int) bool {
+		return h.sortedTS[i].version <= version
+	})
```

Comment:
Ehh, that is not quite right. If you use version as comparator, it is sorted by version, instead of timestamp.

Because the larger the version is, the newer the timestamp will be. So your code works. But it is not same as your thought, maybe.

If you assume the relation between `version` and `timestamp`, I think you could actually just use:

```
cache []struct{
  timestamp int64
  infoschema Is
}
```

And all you need to do is update minimal timestamp at L190.

Reply:
i see, i think search by version should be fine, because it assumes timestamp and version has the same order, it could be used to search both version and timestamp

but you are right, the explicit map is not required and associate timestamp with  infoschema is much cleaner, updating the code